### ISSUE TITLE
"Elastic scrolling" doesn't work

### ISSUE THREAD

**Issue by tajmahal2718** [HexFiend/HexFiend#258]
I know this is probably a minor cosmetic issue only, but it still feels weird that the "elastic scrolling" behaviour introduced with Mac OS X 10.7 Lion doesn't work in Hex Fiend, when it does in basically every other well-behaved native Cocoa application.

This is the rubberband-like "overscrolling" effect that allow you to scroll past the end of the document, which originated from iOS.

Also, I noticed that the overall scrolling behaviour is slightly different from other standard Cocoa views:
- The scroll bar is always present, ignoring the "Show scroll bars" setting in System Preferences, or the fact that the document is too short to require a scroll bar.
- The acceleration that occurs if you rapidly keep scrolling a large scroll view (like the list of mails in Mail.app) with e.g. the trackpad seems to be absent in Hex Fiend.

**Comment by kainjow** [comment]
It's a feature of NSScrollView, but HF uses NSScroller directly.

I think it's non-trivial to rework the UI to use a scroll view instead. I can look into it though.

**Comment by tajmahal2718** [comment]
Would be interesting to know if this works. I recently tried to make an NSTableView with 2 billion rows, and while it basically worked and performance was just fine, the selection highlighting was kind of broken. Probably some kind of rounding error in the NSTableView drawing code with those very large row numbers.

**Comment by kainjow** [comment]
Well, the view sizes are in CGFloat which is a double, so it's all about only rendering what's visible.

With a table, make sure you use things like `noteNumberOfRowsChanged` and `reloadDataForRowIndexes:columnIndexes:` and only load in your data for when the delegates call them.

Similarly with HF, we should be able to make a huge view, and embedding it into a scrollview will cause drawRect: to only be called on the visible frame, but it appears HF isn't fully prepared for that.

**Comment by tajmahal2718** [comment]
Yeah, in theory, it should be fine. But for my quick table view test, I didn't load any data at all. I just returned [NSNumber numberWithInt:rowIndex] as the row value. No changing of data at all either. It's just the bare row (de-)selection mechanism that doesn't work properly as soon as you approach row numbers greater than about 110.000.000. (Also, I get a weird "This NSLayoutConstraint is being configured with a constant that exceeds internal limits." debug message if I have numberOfRowsInTableView: return a number greater than about 500.000.000.) So the problem definitely is somewhere in AppKit's drawing code. (I used a cell-based table view, btw. Haven't checked yet if using a view-based one would make a difference.) Hopefully, it's "only" a problem of NSTableView, and not one of NSScrollView.

So if you go the route of making a huge view, you definitely need to make sure not to hit some possible internal AppKit limits.

**Comment by kainjow** [comment]
Ah, maybe limits changed when they added auto layout.

**Comment by tajmahal2718** [comment]
I actually created this table view test project several years ago, without Auto Layout, so I do not remember if the debug message was already present at that time. But the buggy selection behaviour sure was. (Btw, it seems like the "disable Auto Layout" toggle in IB has disappeared in recent versions of Xcode, but the autoresizing mask in my old nib apparently still works properly. Maybe the mask implicitly gets converted to Auto Layout constraints?)

**Comment by kainjow** [comment]
Trying some things out in https://github.com/ridiculousfish/HexFiend/compare/scrollview-experiments

While I have the views set up, the line numbers are not working correctly. And with even a 10MB file I'm seeing bad performance when trying to drag the scrollbar.

<img width="516" alt="Screen Shot 2020-06-01 at 12 23 35 AM" src="https://user-images.githubusercontent.com/271035/83385609-6998c000-a39e-11ea-8b8e-83bc3f4faee6.png">

I also realized if you change the scroller type to always show in Sys Prefs > General, it will need to expand the window width by the scroller width. Currently it cuts off the text:

<img width="516" alt="Screen Shot 2020-06-01 at 12 27 54 AM" src="https://user-images.githubusercontent.com/271035/83385852-d14f0b00-a39e-11ea-9291-f05f6c4a790d.png">

**Comment by kainjow** [comment]
Closing this out. ROI is low. It's a lot of refactoring and retesting that is very time consuming and expensive (underlying graphic changes require testing on older macOS versions, which is also difficult to do).

**Comment by kainjow** [comment]
I’m reopening this because the way that scrolling works has bothered me. My assessment of the work is the same, but it would improve the native feel of the app, which is important. Also, manually adding scrollers is a fairly legacy way of programming, and you’ll notice there’s no equivalent control in iOS or SwiftUI. It harkens to the days of using scrollers as sliders [example](https://en.m.wikipedia.org/wiki/Scrapbook_(Mac_OS)).

The approach I’d like to take is writing a new control from scratch, that would be a new representer type. It could be enabled via a user default flag. Possible it could show up alongside the current views, instead of replacing them. This would use NSScrollView, and the new control would be the document view.

I’ve been researching other ways of doing this, such as NSCollectionView, but this seems too complex and lots of code rewrite. Lazy grids from SwiftUI seems ideal but has some large memory leaks and would require rewriting almost all the code. It’s unclear if all the supported features can even be done with SwiftUI.

With NSScrollView work can be done incrementally. Chunks of the existing code that would be reused could be refactored out. It’d basically be a combination of the line number view and text views.
